simulation/model.py

```diff
@@ -0,0 +1,328 @@
+from typing import *
+
+from simpy import Environment
+
+from simulation.utils import binary_search, differs_from, AgentId, def_list, merge_sorted
+
+POS_ROUND_DIGITS = 3
+SOFT_COLLIDE_SHIFT = 0.2
+
+
+class CollisionException(Exception):
+    """
+    Thrown when objects on conveyor collide
+    """
+
+    def __init__(self, args):
+        super().__init__('[{}] Objects {} and {} collided in position {}'
+                         .format(args[3], args[0], args[1], args[2]))
+        self.obj1 = args[0]
+        self.obj2 = args[1]
+        self.pos = args[2]
+        self.handler = args[3]
+
+
+class AutomataException(Exception):
+    pass
+
+
+def search_pos(ls: List[Tuple[Any, float]], pos: float,
+               return_index: bool = False,
+               preference: str = 'nearest') -> Tuple[Any, float]:
+    assert len(ls) > 0, "what are you gonna find pal?!"
+    return binary_search(ls, differs_from(pos, using=lambda p: p[1]),
+                         return_index=return_index, preference=preference)
+
+
+def shift(objs, d):
+    return [(o, round(p + d, POS_ROUND_DIGITS)) for (o, p) in objs]
+
+
+# Automata for simplifying the modelling of objects movement
+_model_automata = {
+    'pristine': {'resume': 'moving', 'change': 'dirty'},
+    'moving': {'pause': 'pristine'},
+    'dirty': {'change': 'dirty', 'start_resolving': 'resolving'},
+    'resolving': {'change': 'resolving', 'end_resolving': 'resolved'},
+    'resolved': {'resume': 'moving'}
+}
+
+
+class ConveyorModel:
+    """
+    Datatype which allows for modeling the conveyor with
+    objects moving on top of it.
+
+    A conveyor is modeled as a line with some checkpoints
+    placed along it. Only one checkpoint can occupy a given position.
+
+    Objects can be placed on a conveyor. All the objects on a conveyor
+    move with the same speed - the current speed of a conveyor. Current speed
+    of a conveyor can be changed.
+
+    We want to have the following operations:
+    - put an object to a conveyor (checking for collisions)
+    - remove an object from a conveyor
+    - change the conveyor speed
+    - update the positions of objects as if time T has passed with a current
+      conveyor speed (T might be negative, moving objects backwards)
+    - calculate time T until earliest event of object reaching a checkpoint
+      (or the end of the conveyor), and also return those checkpoint and object
+    """
+
+    def __init__(self, env: Environment, length: float,
+                 checkpoints: List[Tuple[Any, float]],
+                 model_id: AgentId = ('world', 0)):
+        assert length > 0, "Conveyor length <= 0!"
+
+        checkpoints = sorted(checkpoints, key=lambda p: p[1])
```

Comment:
Here it's unlcear what `p[1]` is. I think it is better to have `Dict` instead of `Tuple` and then access to field by it's name which is (I assume) either `timestamp` or `position`.